File: Python/removeSelfReferingFromImportList.py
# -*- coding: utf-8 -*-
# --------------------------------------------------------------------
# Name: removeSelfReferingFromImportList.py
# Removes self refering entries from data import list
# --------------------------------------------------------------------

import arcpy
import os
import subprocess
import sys

provDat = arcpy.GetParameterAsText(0)
ModelDataFolderName = arcpy.GetParameterAsText(1)
pyScript = sys.argv[0]

# execute R-Script
thisScriptPath = os.path.dirname(pyScript)
rScriptPath = os.path.join(os.path.abspath(os.path.join(thisScriptPath, os.pardir)), "R")
rScript = os.path.join(rScriptPath, "removeSelfReferingFromImportList.r")
arcpy.SetProgressor("default", "Executing R Script...")
args = ["R", "--slave", "--vanilla", "--args", provDat, ModelDataFolderName, rScriptPath] 
scriptSource = open(rScript, 'r')
rCommand = subprocess.Popen(args, stdin = scriptSource, stdout = subprocess.PIPE, stderr = subprocess.PIPE, shell = True)
outString, errString = rCommand.communicate()
scriptSource.close()
if errString and "...completed execution of R-script" not in outString:
    arcpy.AddMessage(errString)

# send parameter to ArcGIS
arcpy.SetParameterAsText(2, provDat)

File: R/removeSelfReferingFromImportList.R
# removeSelfReferingFromImportList.r

# Removes self refering entries from data import list

# get input arguments
args <- commandArgs()
provDat <- args[5]
ModelDataFolderName <- args[6]
rScript <- args[7]

# read data import list
csvPath <- paste(provDat, "\\", ModelDataFolderName, "\\use_spatialImport.csv", sep = "")
dat <- read.csv2(csvPath, as.is = T)

# go through all lines
targetPath <- paste(provDat, ModelDataFolderName, sep = "\\")
targetPath <- gsub(targetPath, pattern = "\\\\", replacement = "/")
self <- c()
for (i in 1:nrow(dat)){
  if (is.na(pmatch(targetPath, dat[i,"outPath"])) == F) self <- c(self, i)
}
if (length(self) > 0){
  out <- dat[-self,]
  
  # write to file
  write.table(out, file = csvPath, sep = ";", dec = ".", row.name = F)

}
  
# report to ArcGIS
print ("...finished execution of R-Script.")
